compiler/src/dotty/tools/dotc/transform/init/Objects.scala

```diff
@@ -0,0 +1,1418 @@
+ *
+ *  5. The check is modular in the sense that each object is checked separately and there is no
+ *     whole-program analysis. However, the check is not modular in terms of project boundaries.
+ *
+ */
+object Objects:
+
+  // ----------------------------- abstract domain -----------------------------
+
+  sealed abstract class Value:
+    def show(using Context): String
+
+
+  /**
+   * A reference caches the values for outers and immutable fields.
+   */
+  sealed abstract class Ref(
+    valsMap: mutable.Map[Symbol, Value],
+    varsMap: mutable.Map[Symbol, Heap.Addr],
+    outersMap: mutable.Map[ClassSymbol, Value])
+  extends Value:
+    protected val vals: mutable.Map[Symbol, Value] = valsMap
+    protected val vars: mutable.Map[Symbol, Heap.Addr] = varsMap
+    protected val outers: mutable.Map[ClassSymbol, Value] = outersMap
+
+    def isObjectRef: Boolean = this.isInstanceOf[ObjectRef]
+
+    def klass: ClassSymbol
+
+    def valValue(sym: Symbol): Value = vals(sym)
+
+    def varAddr(sym: Symbol): Heap.Addr = vars(sym)
+
+    def outerValue(cls: ClassSymbol): Value = outers(cls)
+
+    def hasVal(sym: Symbol): Boolean = vals.contains(sym)
+
+    def hasVar(sym: Symbol): Boolean = vars.contains(sym)
+
+    def hasOuter(cls: ClassSymbol): Boolean = outers.contains(cls)
+
+    def initVal(field: Symbol, value: Value)(using Context) = log("Initialize " + field.show + " = " + value + " for " + this, printer) {
+      assert(!field.is(Flags.Mutable), "Field is mutable: " + field.show)
+      assert(!vals.contains(field), "Field already set: " + field.show)
+      vals(field) = value
+    }
+
+    def initVar(field: Symbol, addr: Heap.Addr)(using Context) = log("Initialize " + field.show + " = " + addr + " for " + this, printer) {
+      assert(field.is(Flags.Mutable), "Field is not mutable: " + field.show)
+      assert(!vars.contains(field), "Field already set: " + field.show)
+      vars(field) = addr
+    }
+
+    def initOuter(cls: ClassSymbol, value: Value)(using Context) = log("Initialize outer " + cls.show + " = " + value + " for " + this, printer) {
+      assert(!outers.contains(cls), "Outer already set: " + cls)
+      outers(cls) = value
+    }
+
+  /** A reference to a static object */
+  case class ObjectRef(klass: ClassSymbol)
+  extends Ref(valsMap = mutable.Map.empty, varsMap = mutable.Map.empty, outersMap = mutable.Map.empty):
+    val owner = klass
+
+    def show(using Context) = "ObjectRef(" + klass.show + ")"
+
+  /**
+   * Represents values that are instances of the specified class.
+   *
+   * Note that the 2nd parameter block does not take part in the definition of equality.
+   */
+  case class OfClass private (
+    klass: ClassSymbol, outer: Value, ctor: Symbol, args: List[Value], env: Env.Data)(
+    valsMap: mutable.Map[Symbol, Value], varsMap: mutable.Map[Symbol, Heap.Addr], outersMap: mutable.Map[ClassSymbol, Value])
+  extends Ref(valsMap, varsMap, outersMap):
```

Comment:
This corresponds to `Instance` as defined in Section 4.4. Again, the maps are mutable. We also store its local environment, which stores the information of the local method it resides in. We do not need to store local environment for `ObjectRef` since they refer to global objects.